man/makehelp.1
.TH makehelp 1
.SH NAME
makehelp \- Extract and print help information from Makefiles.
.SH SYNOPSIS
.B makehelp
[\fIOPTION\fR]... [\fIMAKEFILE\fR]...
.SH DESCRIPTION
.B makehelp
extracts and prints help information form Makefiles by looking for documentation comment blocks in front of variables and goals.
A documentation comment block is one or more lines of comments only of which the first line starts with "##".
.br
If no
.IR MAKEFILE
is given,
.B makehelp
tries to read a Makefile from
.IR stdin .
.SH OPTIONS
.TP
.BR \-c ", " \-\-changelog
Prints the changelog and exits.
.TP
.BR \-d ", " \-\-vardefs
Generates variable definition file and exits.
.TP
.BR \-h ", " \-\-help
Prints this help text and exits.
.TP
.BR \-?
Same as -h.
.TP
.BR \-V ", " \-\-version
Prints version information and exits.
.SH MAKEHELP COMMENT SYNTAX
.SS Documenting Variables
In order to document a variable, use a documentation comment before the variable declaration.
It is common to document those variables which the user could or should override on the command line.

.RS 4
## C Compiler executable.
.RS 0
CC:=gcc
.SS Documenting Goals
In order to document a goal, use a documentation comment before the goal declaration.
It is common to document those goals which are relevant to the user.

.RS 4
\[char46]PHONY: all
.RS 0
## Builds the executable.
.RS 0
all: main
.SS Referencing Variables
It is possible to reference variables in comments.
To do so, include a variable in a comment like this: $(VARNAME).

.RS 4
## Installation path prefix.
.RS 0
# Current value: $(PREFIX).
.RS 0
PREFIX:=/usr/local/
.SH EXIT STATUS
The exit status is 0 if help text could be printed.
If an error occurred, the exit status is not zero.
.SH AUTHOR
Written by Christian Hujer.
.SH REPORTING BUGS
Report bugs via <https://github.com/christianhujer/makehelp/issues>
.SH COPYRIGHT
Copyright © 2015 Chrsitian Hujer.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


makehelp.pl
#!/usr/bin/perl
# Prints help for a Makefile.
# Usual invocation (from make):
# help:
# 	perl makehelp.pl $(MAKEFILE_LIST)
#
# To make your make inline help documentation work, start documentation comments before a target or variable with double-hash.
# Example:
# .PHONY: all
# ## Compiles and links all sources.
# # This is the default target.
# all: ...

$version = "1.3";

sub changelog() {
    print <<END;
2014-07-07 1.3
    Make the command line options more similar to glibc argp.

2012-11-11 1.2
    Implement changelog option.

END
}

sub help() {
    print <<END;
Usage: perl $0 MAKEFILE...
Prints the help text of one or more MAKEFILEs.

To be invoked from the Makefile with a rule like this:
help:
	perl $0 \$(MAKEFILE_LIST)

Supported options:
  -c, --changelog   Print the changelog.
  -?, --help        Print this help text.
  -h                Same as -?
  -V, --version     Print version information.

$0 prints the documentation of variables and goals.
A documentation comment is recognized as one or more comment lines right before a goal or variable definition,
of which the first comment line starts with ##.
The documentation comment may contain references to environment variables.
All references to environment variables will be replaced with their values.
This includes the make goal.
Note: To reference variables which are defined in the Makefile, don't forget to export them.

Example:
  .PHONY: all
  ## Performs the complete build of the project.
  # This is the default target.
  all: mybinary

  ## Flags for the C preprocessor.
  CPPFLAGS?=

  .PHONY: help
  ## Prints this help text.
  help:
  	perl $0 \$(MAKEFILE_LIST)

Report bugs to cher\@riedquat.de.
END
}

sub version() {
    print <<END;
makehelp.pl $version
(c) 2011 - 2012 Christian Hujer. All rights reserved.
Licensed under GPLv3.
See http://www.gnu.org/licenses/gpl.html for license information.
END
}

while ($_ = shift) {
    /^-(c|-?changelog)$/ and changelog() and exit 0;
    /^-(\?|h|-?help)$/ and help() and exit 0;
    /^-(V|-?version)$/ and version() and exit 0;
    /^(-.*)$/ and die "$0: Unknown command line option $1. Try $0 --help.";
    push @RealArgs, $_;
}
@ARGV = @RealArgs;

while (<>) {
    if (/^## .*$/ ... /^([^.#\t][^:=]*:([^=]|$)|([^.#\t][^:=?]*(\?=|:=|=))|$)/) {
        s/^##? ?//;
        $h .= $p;
        $p = $_;
    } elsif ($h) {
        $p =~ s/(\?=|:=|=).*/=/;
        $p =~ s/:.*/:/;
        chomp $p;
        $p =~ s/\$\(([^)]*)\)/$ENV{$1}/ge;
        $h =~ s/\$\(([^)]*)\)/$ENV{$1}/ge;
        if ($p =~ /=$/) {
            $p =~ s/=$//;
            if ($h ne $vars{$p}) {
                $vars{$p} .= $h;
            }
        } else {
            $p =~ s/:$//;
            if ($h ne $goals{$p}) {
                $goals{$p} .= $h;
            }
        }
        undef $h;
        undef $p;
    }
}

print <<END;
Usage: make [OPTION|GOAL|VARIABLE]...
Runs make to make the specified GOALs.
If no GOAL is specified, the default goal is made (usually "all").

Popular make OPTIONs:
  -s    Silent mode, disables command echo.
  -k    Keep going, continues after errors.
  -j N  Run N jobs in parallel.
  -n    Don't run the commands, just print them.
  -q    Run no commands; exit status says if up to date.
  -h    Print make help text.
Use option -h to lists the GNUmake part of the help.
END

my $indent = " " x 20;
if (%vars) {
    print <<END;

A VARIABLE is specified as name=value pair.
Supported VARIABLEs:
END

    for (sort keys %vars) {
        $vars{$_} =~ s/^/$indent/gm;
        if (length > 16) {
            print "  $_\n$vars{$_}";
        } else {
            $tmpIndent = " " x (2 + length);
            $vars{$_} =~ s/^$tmpIndent//;
            print "  $_$vars{$_}";
        }
    }
}

print <<END;

Supported GOALs:
END

for (sort keys %goals) {
    $goals{$_} =~ s/^/$indent/gm;
    if (length > 16) {
        print "  $_\n$goals{$_}";
    } else {
        $tmpIndent = " " x (2 + length);
        $goals{$_} =~ s/^$tmpIndent//;
        print "  $_$goals{$_}";
    }
}
